rust/libnewsboat/src/configcontainer.rs

```diff
@@ -0,0 +1,1360 @@
+            default_value: val.to_string(),
+            data_type: ConfigDataType::Enum,
+            enum_values: values.iter().map(|s| s.to_string()).collect(),
+            multi_option: false,
+        }
+    }
+
+    pub fn set_value(&mut self, new_val: String) -> Result<(), String> {
+        match self.data_type {
+            ConfigDataType::Bool => {
+                if new_val == "true" || new_val == "yes" || new_val == "false" || new_val == "no" {
+                    self.value = new_val;
+                    Ok(())
+                } else {
+                    Err(format!("invalid boolean value: {new_val}"))
+                }
+            }
+            ConfigDataType::Int => {
+                if new_val.parse::<isize>().is_ok() {
+                    self.value = new_val;
+                    Ok(())
+                } else {
+                    Err(format!("invalid integer value: {new_val}"))
+                }
+            }
+            ConfigDataType::Enum => {
+                if self.enum_values.contains(&new_val) {
+                    self.value = new_val;
+                    Ok(())
+                } else {
+                    Err(format!("invalid enum value: {new_val}"))
+                }
+            }
+            _ => {
+                self.value = new_val;
+                Ok(())
+            }
+        }
+    }
+}
+
+pub struct ConfigContainer {
+    config_data: Arc<Mutex<BTreeMap<String, ConfigData>>>,
+}
+
+impl Default for ConfigContainer {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl ConfigContainer {
+    pub fn new() -> Self {
+        let mut config_data = BTreeMap::new();
+
+        config_data.insert(
+            "always-display-description".to_string(),
+            ConfigData::new("false", ConfigDataType::Bool),
+        );
+
+        config_data.insert(
+            "article-sort-order".to_string(),
+            ConfigData::new("date-asc", ConfigDataType::Str),
+        );
+
+        config_data.insert(
+            "articlelist-format".to_string(),
+            ConfigData::new("%4i %f %D %6L  %?T?|%-17T|  &?%t", ConfigDataType::Str),
+        );
+
+        config_data.insert(
+            "auto-reload".to_string(),
+            ConfigData::new("no", ConfigDataType::Bool),
+        );
+
+        config_data.insert(
+            "bookmark-autopilot".to_string(),
+            ConfigData::new("false", ConfigDataType::Bool),
+        );
+
+        config_data.insert(
+            "bookmark-cmd".to_string(),
+            ConfigData::new("", ConfigDataType::Str),
+        );
+
+        config_data.insert(
+            "bookmark-interactive".to_string(),
+            ConfigData::new("false", ConfigDataType::Bool),
+        );
+
+        let default_browser = crate::utils::get_default_browser();
+        config_data.insert(
+            "browser".to_string(),
+            ConfigData::new(&default_browser.to_string_lossy(), ConfigDataType::Path),
+        );
```

Comment:
Converting `Filepath` to `String` enforces UTF-8 on it, and might replace some characters with Unicode's "replacement character". This is bad because the browser won't work.

In C++, we dodged this by treating `std::string` as a container of bytes, which is then interpreted as `Filepath` or `String` depending on what method is querying it. We could do the same in Rust if we used `Vec<u8>`, but I'd prefer to add two new `Filepath` fields to `ConfigData` and store correctly typed values there. Getters will have to be adjusted to fetch from the correct fields. If you want a Rust exercise, you can then turn `ConfigData` into an `enum` so only the relevant fields are available; that's for a separate PR though.

What do you think of this?